File: frontend/src/types/llmSettings.ts
export interface LLMSettingsUpdate {
  chat_base_url?: string | null
  chat_api_key?: string | null
  chat_model?: string | null
  chat_temperature?: number | null

  embedding_base_url?: string | null
  embedding_api_key?: string | null
  embedding_model?: string | null

  image_base_url?: string | null
  image_api_key?: string | null
  image_model?: string | null
}

export interface LLMSettingsDefaults {
  chat_base_url: string
  chat_model: string
  embedding_base_url: string
  embedding_model: string
  image_base_url: string
  image_model: string
}

export interface LLMSettingsResponse {
  chat_base_url: string | null
  chat_api_key_set: boolean
  chat_model: string | null
  chat_temperature: number | null

  embedding_base_url: string | null
  embedding_api_key_set: boolean
  embedding_model: string | null

  image_base_url: string | null
  image_api_key_set: boolean
  image_model: string | null

  defaults: LLMSettingsDefaults | null
}

export interface MaskedKeys {
  chat_api_key: string
  embedding_api_key: string
  image_api_key: string
}

File: backend/app/models/llm_settings.py
from datetime import datetime

from sqlalchemy import Column, DateTime, Float, Integer, String, Text

from app.db.base import Base


class LLMSettings(Base):
    """Per-user LLM configuration overrides.

    Each row stores one user's customized LLM provider settings.
    Fields left empty/null fall back to the server-wide env defaults.
    """

    __tablename__ = "llm_settings"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, unique=True, nullable=False, index=True)

    # Text generation
    chat_base_url = Column(String(500), nullable=True)
    chat_api_key = Column(Text, nullable=True)
    chat_model = Column(String(255), nullable=True)
    chat_temperature = Column(Float, nullable=True)

    # Embedding
    embedding_base_url = Column(String(500), nullable=True)
    embedding_api_key = Column(Text, nullable=True)
    embedding_model = Column(String(255), nullable=True)

    # Image generation
    image_base_url = Column(String(500), nullable=True)
    image_api_key = Column(Text, nullable=True)
    image_model = Column(String(255), nullable=True)

    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )
